File: javascript/Tournament.js
//
// This is only a SKELETON file for the 'Tournament' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const tournamentTally = (results) => {
    const table = ['Team                           | MP |  W |  D |  L |  P']
    if (!results) {
        return table.join('\n')
    }

    const teams = new Map()
    for (const result of results.split('\n')) {
        const [teamA, teamB, outcome] = result.split(';')
        if (!teams.get(teamA)) {
            teams.set(teamA, { win: 0, draw: 0, loss: 0 })
        }
        if (!teams.get(teamB)) {
            teams.set(teamB, { win: 0, draw: 0, loss: 0 })
        }

        switch (outcome) {
            case 'win':
                teams.get(teamA).win++;
                teams.get(teamB).loss++;
                break
            case 'draw':
                teams.get(teamA).draw++;
                teams.get(teamB).draw++;
                break
            case 'loss':
                teams.get(teamA).loss++;
                teams.get(teamB).win++
        }
    }

    for (const [team, { win, draw, loss }] of sort(teams.entries())) {
        table.push(
            `${team.padEnd(30)} | ${String(win + draw + loss).padStart(2)} | ${String(win).padStart(2)} | ${String(draw).padStart(2)} |` +
            ` ${String(loss).padStart(2)} | ${String(3 * win + draw).padStart(2)}`
        )
    }

    return table.join('\n')
}

function sort(entries) {
    return [...entries].sort(([aName, a], [bName, b]) => {
        const aPoints = 3 * a.win + a.draw,
            bPoints = 3 * b.win + b.draw
        if (aPoints === bPoints) {
            return aName.localeCompare(bName)
        }

        return bPoints - aPoints
    })
}

File: python/Tournament.py
from operator import itemgetter
from typing import Counter, DefaultDict, List, Sequence, Tuple, Union

ROW_FORMAT = "{:<30} | {:>2} | {:>2} | {:>2} | {:>2} | {:>2}"
OUTCOME_MAP = {"win": "loss", "loss": "win", "draw": "draw"}


Cell = Union[str, int]
Row = Tuple[str, Cell, Cell, Cell, Cell, Cell]

def tally(results: Sequence[str]) -> List[str]:
    """
    Tally a football tournament.
    """
    teams: DefaultDict[str, Counter[str]] = DefaultDict(Counter)
    for result in results:
        home, away, outcome = result.split(";")
        teams[home][outcome] += 1
        teams[away][OUTCOME_MAP[outcome]] += 1

    table: List[Row] = []
    for team, record in sorted(teams.items()):
        wins, draws, losses = record["win"], record["draw"], record["loss"]
        matches, points = wins + draws + losses, 3 * wins + draws
        table.append((team, matches, wins, draws, losses, points))
    table.sort(key=itemgetter(-1), reverse=True)

    table.insert(0, ("Team", "MP", "W", "D", "L", "P"))
    return [ROW_FORMAT.format(*row) for row in table]